### PR TITLE
cql3,index: add panic checks to base key generation

### PR DESCRIPTION
**Pull request by psarna**
In order to be extra sure that we always generate proper
base partition/clustering keys from paging info when executing
an indexed query, additional checks are added - if any of them
triggers, an exception will be thrown.
Created in order to help debug an existing issue:
Refs #5856

Tests: unit(dev)

### CODE REVIEW

File: cql3/statements/select_statement.cc [hunk 1]
```diff
@@ -439,9 +439,25 @@ generate_base_key_from_index_pk(const partition_key& index_pk, const std::option
...
+            throw std::runtime_error(format("Base key column not found in the view: {}", base_col.name_as_text()));
+        }
+        if (base_col.type != view_col->type) {
+            throw std::runtime_error(format("Mismatched types for base and view columns {}: {} and {}",
+                    base_col.name_as_text(), base_col.type->cql3_type_name(), view_col->type->cql3_type_name()));
+        }
         if (view_col->is_partition_key()) {
             exploded_base_key.push_back(index_pk.get_component(view_schema, view_col->id));
-        } else if (index_ck) {
+        } else  {
```
Comment (penberg) [inline cql3/statements/select_statement.cc:451]: Extra space before `{`.

**Comment by nyh** [comment]
@psarna you refer to https://github.com/scylladb/scylla/issues/6087, but this is some closed LWT issue. Maybe you have a typo in the issue number?

**Comment by psarna** [comment]
Indeed, I updated the message, thanks.

**Comment by nyh** [comment]
Can you please modify the patch commit message itself? Because this is a single-patch series, I'm not going to merge the cover letter anyway, so I care more about the message in the actual commit. Thanks.

File: cql3/statements/select_statement.cc [hunk 2]
```diff
@@ -439,9 +439,25 @@ generate_base_key_from_index_pk(const partition_key& index_pk, const std::option
...
+            throw std::runtime_error(format("Mismatched types for base and view columns {}: {} and {}",
+                    base_col.name_as_text(), base_col.type->cql3_type_name(), view_col->type->cql3_type_name()));
+        }
         if (view_col->is_partition_key()) {
             exploded_base_key.push_back(index_pk.get_component(view_schema, view_col->id));
-        } else if (index_ck) {
+        } else {
+            if (!view_col->is_clustering_key()) {
+                throw std::runtime_error(
+                        format("Base primary key column is not a primary key column in the materialized view: {} (kind: {})",
```
Comment (bhalevy) [inline cql3/statements/select_statement.cc:454]: s/primary/clustering/?

Comment (psarna) [reply cql3/statements/select_statement.cc:454]: Both are correct, will be unique in the logs, and are only expected to be triggered by a bug in Scylla - so I think the current message is also fine. If you prefer to switch to `clustering`, let me know and I'll update the patch.

**Comment by psarna** [comment]
> Can you please modify the patch commit message itself? Because this is a single-patch series, I'm not going to merge the cover letter anyway, so I care more about the message in the actual commit. Thanks.

done

File: cql3/statements/select_statement.cc (hunk 2 above)
Comment (nyh) [reply cql3/statements/select_statement.cc:454]: The message is correct because the first if checked is_partition_key() and the second is_clustering_key(). If we get to print this message, it means it is neither.

File: cql3/statements/select_statement.cc [hunk 3]
```diff
@@ -439,9 +439,25 @@ generate_base_key_from_index_pk(const partition_key& index_pk, const std::option
...
+                                view_col->name_as_text(), to_sstring(view_col->kind)));
+            }
+            if (!index_ck) {
+                throw std::runtime_error(format("Base primary key column is not present in the materialized view key: {}",
+                        view_col->name_as_text()));
```
Comment (nyh) [inline cql3/statements/select_statement.cc:459]: I can't make sense of this check, can you please verify and/or explain it to me? How does "index_ck" (what is it?) not existing related to the message you print, "base primary key column not present", and if it's base column, why do you print view_col)?

File: cql3/statements/select_statement.cc (hunk 2 above)
Comment (bhalevy) [reply cql3/statements/select_statement.cc:454]: hmm, it seems confusing to me because you check for `!view_col->is_clustering_key()`

File: cql3/statements/select_statement.cc (hunk 3 above)
Comment (psarna) [reply cql3/statements/select_statement.cc:459]: I'll try to rephrase this message. What I wanted to express is that we reached the following situation:
* we want to get the value which corresponds to the base key column `base_col` - it should either be part of the index partition key or its clustering key
* we already determined that it's not part of the index partition key
* but, now we see that the paging state from the indexed query, which is supposed to store index's partition and clustering keys, does not have a clustering key after all! Which means that somehow the column that corresponds to the base column `base_col` was not shipped to us in the paging state, which must be a bug

To sum up, by `base column ${NAME}` I really mean `a column in the index, which corresponds to a base column with given name ${NAME}`. It's not easy to express that in a single small error message :)

File: cql3/statements/select_statement.cc (hunk 2 above)
Comment (nyh) [reply cql3/statements/select_statement.cc:454]: But it follows the other if(view_col->is_partition_key())), if that other if() succeeded we wouldn't even get here...

Comment (psarna) [reply cql3/statements/select_statement.cc:454]: @bhalevy only because it's the `else` branch of `view_col->is_partition_key()`, so it's already implied that it wasn't found in the partition key

**Comment by psarna** [comment]
I rephrased the error messages

**Comment by nyh** [comment]
Committed to next. It's a single patch, not a merge, so github will never know so I'm closing manually....

File: cql3/statements/select_statement.cc (hunk 2 above)
Comment (bhalevy) [reply cql3/statements/select_statement.cc]: thanks
